README
Qore openldap module

INTRODUCTION
------------
The Qore openldap module provides LDAP functionality to Qore by using the
openldap (http://www.openldap.org) client library to provide a Qore API to
talk to LDAP servers.

The LDAP API is standard but this module was specifically designed to compile
and link with the openldap APIs and has not been tested with other LDAP APIs.

see docs for more information


BUILDING
--------
Requires qore 0.8.7+ and openssl headers and libraries to build and run

to configure the build, execute
        configure --disable-debug

if the qore library cannot be found, then you can use:
        configure --disable-debug --with-qore=<dir>

if the openldap library or headers cannot be found, then you can use:
        configure --disable-debug --with-openldap=<dir>

(or set the OPENLDAP_DIR environment variable before callinf configure)

The qore binary also needs to be in the path so configure can determine the module directory

Then execute

make && make install

(or 'make && sudo make install' as needed)

please direct any questions to:
david@qore.org


ChangeLog
2013-02-04 08:57  david_nichols

	* test/ldapmodify, test/ldapsearch, test/qldapmodify,
	  test/qldapsearch: updated test programs with q prefix and -W
	  support

2012-12-28 19:59  david_nichols

	* src/QoreLdapClient.h: updated password check in bind function

2012-12-28 08:32  david_nichols

	* Makefile.am, docs/mainpage.doxygen.tmpl, src/QC_LdapClient.qpp,
	  src/QoreLdapClient.h, test/ldapadd, test/ldapdelete,
	  test/qldapadd, test/qldapdelete, test/qldappasswd: added the
	  LdapClient::passwd() method to change the ldap password, added
	  qldappasswd example program; renamed example programs to be
	  prefixed with a "q" to not conflict with existing standard ldap
	  programs (mostly because these programs are not compatible with
	  the standard commands), implemented support for prompting for
	  passwords in all example programs and not echoing the password
	  using TermIOS functionality

2012-12-26 08:53  david_nichols

	* docs/mainpage.doxygen.tmpl, src/QC_LdapClient.qpp: doc
	  corrections/updates

2012-12-23 08:28  david_nichols

	* docs/mainpage.doxygen.tmpl: doc updates

2012-12-21 16:36  david_nichols

	* RELEASE-NOTES, configure.ac, docs/mainpage.doxygen.tmpl,
	  qore-openldap-module.spec, src/QC_LdapClient.qpp,
	  src/QoreLdapClient.h: added the LdapClient::rename() method;
	  updated the module's version to 1.0 for the initial release

2012-12-21 16:02  david_nichols

	* src/QC_LdapClient.qpp, src/QoreLdapClient.h: implemented
	  LdapClient::compare()

2012-12-21 13:44  david_nichols

	* Makefile.am, qore-openldap-module.spec: rpm build fixes

2012-12-21 13:17  david_nichols

	* docs/doxyfile.tmpl, docs/mainpage.doxygen.tmpl,
	  src/QC_LdapClient.qpp, src/QoreLdapClient.h, test/ldapadd,
	  test/ldapdelete, test/ldapmodify, test/ldapsearch: fixed timeout
	  handling; added optional timeout parameters to all methods
	  requiring server communication, updated docs

2012-12-21 10:24  david_nichols

	* src/QC_LdapClient.qpp, src/QoreLdapClient.h, test/ldapadd,
	  test/ldapdelete, test/ldapmodify, test/ldapsearch: added support
	  for the "starttls" option in the LdapClient::constructor();
	  redesigned initialization and object copying to better maintain
	  options, implemented -Z option in all scripts to ensure a secure
	  connection

2012-12-21 08:57  david_nichols

	* src/QC_LdapClient.qpp, src/QoreLdapClient.h, test/ldapadd,
	  test/ldapdelete, test/ldapmodify: added delete and add
	  functionality, added corresponding scripts, fixed a memory bug in
	  adding lists of attribute values

2012-12-20 13:24  david_nichols

	* src/QoreLdapClient.h: fixed delete / delete [] operator mismatch

2012-12-20 13:16  david_nichols

	* src/QC_LdapClient.qpp, src/QoreLdapClient.h, test/ldapmodify,
	  test/ldapsearch: working LdapClient::modify() method, much
	  improved error handling, bind error handling, verbose error
	  messages from parsed ldap results

2012-12-20 07:14  david_nichols

	* src/QoreLdapClient.h, src/openldap-module.cpp: disabled the
	  QLO_DISABLE_OPENSSL_CLEANUP option checking and setting in
	  openldap module initialization, the crashes when shutting down
	  were caused by an unrelated memory bug (delete / delete []
	  mismatch - caught by valgrind)

2012-12-19 19:12  david_nichols

	* src/QC_LdapClient.qpp, src/QoreLdapClient.h,
	  src/openldap-module.cpp, src/openldap-module.h: started adding
	  infrastructure support for supporting arbitrary ldap modify
	  commands

2012-12-19 16:47  david_nichols

	* src/QoreLdapClient.h, test/ldapsearch: added an option to supress
	  referalls, streamlined option handling in ldapsearch

2012-12-19 14:12  david_nichols

	* src/QoreLdapClient.h, test/ldapsearch: implemented support for
	  setting the ldap protocol level

2012-12-19 13:47  david_nichols

	* src/QC_LdapClient.qpp, src/QoreLdapClient.h, test/ldapsearch:
	  added support for setting the default timeout for all ldap
	  operations

2012-12-19 11:03  david_nichols

	* src/openldap-module.cpp: do not allow the module to be loaded if
	  QLO_DISABLE_OPENSSL_CLEANUP is already enabled; set
	  QLO_DISABLE_OPENSSL_CLEANUP when initializing the module since
	  the openldap library will perform its own cleanup of the openssl
	  library and doing it twice will cause a segfault when shutting
	  down qore

2012-12-12 11:49  david_nichols

	* configure.ac, src/QC_LdapClient.qpp, src/QoreLdapClient.h,
	  test/ldap.q, test/ldapsearch: updated to require qore 0.8.7;
	  added support for asynchronous operations; renamed ldap.q ->
	  ldapsearch

2012-12-12 09:44  david_nichols

	* src/QoreLdapClient.h, test/ldap.q: updates

2012-11-07 16:59  david_nichols

	* src/QC_LdapClient.qpp, test/ldap.q: implemented support for
	  setting the search scope

2012-11-07 14:43  david_nichols

	* src/QC_LdapClient.qpp, src/QoreLdapClient.h, test/ldap.q: added
	  ability specify attributes in a search

2012-11-07 10:22  david_nichols

	* configure.ac: require 0.8.6

2012-11-07 08:13  david_nichols

	* src/QoreLdapClient.h: fixed memory mgmt with LDAPAPIInfo when
	  there are no extensions

2012-11-07 07:55  david_nichols

	* src/QC_LdapClient.qpp, src/QoreLdapClient.h, test/ldap.q:
	  progress on returning ldap search results

2012-11-05 16:19  david_nichols

	* ., AUTHORS, COPYING, ChangeLog, Makefile.am, README,
	  RELEASE-NOTES, configure.ac, docs, docs/doxyfile.tmpl,
	  docs/footer_template.html, docs/mainpage.doxygen.tmpl, m4,
	  m4/acx_pthread.m4, qore-openldap-module.spec, reconf.sh, src,
	  src/Makefile.am, src/QC_LdapClient.qpp, src/QoreLdapClient.h,
	  src/openldap-module.cpp, src/openldap-module.h,
	  src/single-compilation-unit.cpp, test, test/ldap.q: checked in
	  initial version of openldap module



src/openldap-module.h
/* -*- mode: c++; indent-tabs-mode: nil -*- */
/*
  openldap-module.h
  
  Qore Programming Language

  Copyright 2003 - 2010 David Nichols

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#ifndef _QORE_OPENLDAP_MODULE_H
#define _QORE_OPENLDAP_MODULE_H

#include "../config.h"

#include <qore/Qore.h>

#include <ldap.h>

#include <map>
#include <string>

typedef std::map<std::string, int> strintmap_t;

class ModMap : public strintmap_t {
public:
   DLLLOCAL ModMap() {
      insert(ModMap::value_type("add", LDAP_MOD_ADD));
      insert(ModMap::value_type("delete", LDAP_MOD_DELETE));
      insert(ModMap::value_type("replace", LDAP_MOD_REPLACE));
   }

   DLLLOCAL int get(const char* mod) const {
      const ModMap::const_iterator i = find(mod);
      return i == end() ? -1 : i->second;
   }
};

DLLLOCAL extern ModMap modmap;

#endif


src/openldap-module.cpp
/* indent-tabs-mode: nil -*- */
/*
  openldap Qore module

  Copyright (C) 2012 David Nichols, all rights reserved

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
*/

#include "openldap-module.h"

#include "QoreLdapClient.h"

static QoreStringNode *openldap_module_init();
static void openldap_module_ns_init(QoreNamespace *rns, QoreNamespace *qns);
static void openldap_module_delete();

// qore module symbols
DLLEXPORT char qore_module_name[] = "openldap";
DLLEXPORT char qore_module_version[] = PACKAGE_VERSION;
DLLEXPORT char qore_module_description[] = "openldap module";
DLLEXPORT char qore_module_author[] = "David Nichols";
DLLEXPORT char qore_module_url[] = "http://qore.org";
DLLEXPORT int qore_module_api_major = QORE_MODULE_API_MAJOR;
DLLEXPORT int qore_module_api_minor = QORE_MODULE_API_MINOR;
DLLEXPORT qore_module_init_t qore_module_init = openldap_module_init;
DLLEXPORT qore_module_ns_init_t qore_module_ns_init = openldap_module_ns_init;
DLLEXPORT qore_module_delete_t qore_module_delete = openldap_module_delete;
DLLEXPORT qore_license_t qore_module_license = QL_LGPL;

DLLLOCAL QoreClass* initLdapClientClass(QoreNamespace& ns);

// modify action map
ModMap modmap;

static QoreNamespace OLNS("OpenLdap");

static QoreStringNode* openldap_module_init() {
   // this also serves to initialize the library in a single-threaded way
   QoreStringNode* err = QoreLdapClient::checkLibrary();
   if (err)
      return err;

   OLNS.addSystemClass(initLdapClientClass(OLNS));

   return 0;
}

static void openldap_module_ns_init(QoreNamespace* rns, QoreNamespace* qns) {
   qns->addNamespace(OLNS.copy());
}

static void openldap_module_delete() {
}
